packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts

```diff
@@ -110,6 +166,20 @@ export class FederatedQuadSource implements ILazyQuadSource {
     return false;
   }
 
+  /**
+   * Get the unique, deterministic id for the given source.
+   * @param source A data source.
+   * @return The numerical id of the given source.
+   */
+  public getSourceId(source: IDataSource): number {
+    let sourceId = this.sourceIds.get(source);
+    if (sourceId === undefined) {
+      sourceId = this.sourceIds.size;
```

Comment:
Actually, you _might_ want to give sources the opportunity to optionally define their own ID. Then, if a source is hidden behind a cache or something, it would still give the same ID.